### ISSUE TITLE
Heapster need IPv4 to work

### ISSUE THREAD

**Issue by valentin2105** [kubernetes/heapster#1852]
Hi,

I'm trying to use heapster in k8s IPv6 only environment, 
It seem that Heapster try to get the IPv4 IP of the node, but my workers doesn't get any. 

Kubelet is running with `  --node-ip=<IPv6>`

```
E1017 21:26:05.000410       1 kubelet.go:280] Node florida.k8s.domain.com has no valid hostname and/or IP address: florida.k8s.domain.com
E1017 21:26:05.000420       1 kubelet.go:280] Node texas.k8s.domain.com has no valid hostname and/or IP address: texas.k8s.domain.com
E1017 21:26:05.000427       1 kubelet.go:280] Node alabama.k8s.domain.com has no valid hostname and/or IP address: alabama.k8s.domain.com
```

I can't give more information if needed.

**Comment by DirectXMan12** [comment]
hmm... first, seems we always error out if we don't get an IP address (which seems wrong), but you should probably still report a node IP on the node, even with an IP v6 address.  What does `kubectl get node -o yaml` show under the addresses section?

**Comment by valentin2105** [comment]
`kubectl get node -o yaml `

```
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2017-10-16T04:58:50Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: texas.k8s.srv.example.com
    name: texas.k8s.srv.example.com
    namespace: ""
  spec:
    externalID: texas.k8s.srv.example.com
  status:
    addresses:
    - address: 1204:f200:d:f:c8bc:23ff:ffee:f71f
      type: InternalIP
    - address: texas.k8s.srv.example.com
      type: Hostname
    allocatable:
      cpu: "8"
      memory: 8076356Ki
      pods: "110"
    capacity:
      cpu: "8"
      memory: 8178756Ki
      pods: "110"

```

**Comment by DirectXMan12** [comment]
whoops, I missed a line when I was reading over the code last time: https://github.com/kubernetes/heapster/blob/e90c142b41c709bf47403e5756536ac961294f3d/metrics/sources/kubelet/kubelet.go#L323.

We explicitly try to convert to an IPv4 address.  That seems wrong, so we should fix that.  There's at least one other IPv6 issue with Heapster and nodes though (Heapster uses `fmt.Sprintf` to join host and port, which doesn't work for IPv6, since you need to surround the address with square brackets).

We'll need to fix both those issues to get anywhere on IPv6 support.

**Comment by valentin2105** [comment]
Thanks, this PR seem a good way to make Heapster work on IPv6, 
Any idea when a release will be available ? 

I will do a feedback. 😄

**Comment by DirectXMan12** [comment]
Closed by #1861
